src/futures_tuple_set.rs

```diff
@@ -68,11 +65,41 @@ where
 
         Poll::Ready((res, data))
     }
+
+    /// Returns an iterator over all futures of type `T` pushed via [`FuturesTupleSet::try_push`].
+    /// The order that futures are returned is not guaranteed.
+    ///
+    /// If downcasting a future to `T` fails it will be skipped in the iterator.
+    pub fn iter_of_type<T>(&self) -> impl Iterator<Item = (&T, &D)>
+    where
+        T: 'static,
+    {
+        self.inner
+            .iter_of_type()
+            .map(|(id, item)| (item, self.data.get(&id).expect("must have data for future")))
+    }
+
+    /// Returns an iterator with mutable access over all futures of type `T`
+    /// pushed via [`FuturesTupleSet::try_push`].
+    /// The order that futures are returned is not guaranteed.
+    ///
+    /// If downcasting a future to `T` fails it will be skipped in the iterator.
+    pub fn iter_mut_of_type<T>(&mut self) -> impl Iterator<Item = (&mut T, &D)>
+    where
+        T: 'static,
+    {
+        // TODO: work out how to return &mut D here, without a
+        // "captured variable cannot escape `FnMut` closure body" compiler error
+        self.inner
+            .iter_mut_of_type()
+            .map(|(id, item)| (item, self.data.get(id).expect("must have data for future")))
+    }
```

Comment:
I couldn't get this method to return a `&mut D`. I tried a few different ways, and I'd always end up with an error like:
<img width="1148" alt="Screenshot 2025-05-08 at 13 41 03" src="https://github.com/user-attachments/assets/01ae5785-e8c2-4b03-9268-d61fc9679ec8" />

Reply:
Hmm, I'll see if I can make this work tomorrow.

Reply:
Got it working, the solution is to remove the mutable refs from a constructed `HashMap`. This convinces the compiler that we can't return the same mutable ref twice.